Running `kubermatic-installer deploy --migrate-gateway-api` without `migrateGatewayAPI: true` set leads to SIGSEGV

### What happened?

I thought I can first run `kubermatic-installer deploy --migrate-gateway-api` (other parameters omitted) to deploy Envoy Gateway and have the Gateway created. And only in a second step add `migrateGatewayAPI: true` (etc.) to the values.yaml.

However, that's not the case. And what's worse, it leads to kubermatic-installer being stuck for quite a long time, and then running into a SIGSEGV:

```
[...]
time="2026-03-27T11:10:13Z" level=info msg="   📦 Deploying envoy-gateway-controller"
time="2026-03-27T11:10:13Z" level=debug msg="      Ensuring namespace…" namespace=envoy-gateway-controller
time="2026-03-27T11:10:13Z" level=debug msg="      Checking for release…" name=envoy-gateway-controller namespace=envoy-gateway-controller
time="2026-03-27T11:10:13Z" level=debug msg="$ KUBECONFIG=/builds/sck/sck/.kubeconfig helm --namespace envoy-gateway-controller list --all -o json"
time="2026-03-27T11:10:13Z" level=info msg="      Deploying Helm chart..."
time="2026-03-27T11:10:13Z" level=debug msg="      Installing…"
time="2026-03-27T11:10:13Z" level=debug msg="Skipping OCI repository: oci://docker.io/envoyproxy"
time="2026-03-27T11:10:13Z" level=debug msg="$ KUBECONFIG=/builds/sck/sck/.kubeconfig helm --namespace default dependency build /usr/local/lib/charts/envoy-gateway-controller"
time="2026-03-27T11:10:14Z" level=debug msg="$ KUBECONFIG=/builds/sck/sck/.kubeconfig helm --namespace envoy-gateway-controller upgrade --install --timeout 5m0s --values /tmp/helmvalues.2933479062 envoy-gateway-controller /usr/local/lib/charts/envoy-gateway-controller"
time="2026-03-27T11:10:27Z" level=info msg="      Waiting for GatewayClass to be available..."
time="2026-03-27T11:10:47Z" level=info msg="      GatewayClass \"kubermatic-envoy-gateway\" is available."
time="2026-03-27T11:10:47Z" level=info msg="   ✅ Success."
[...]
time="2026-03-27T11:10:51Z" level=info msg="   📡 Determining DNS settings…"
time="2026-03-27T11:10:51Z" level=debug msg="   Waiting for Gateway to be ready" gateway=kubermatic/kubermatic
time="2026-03-27T11:10:51Z" level=debug msg="   failed to get Gateway, err: Gateway.gateway.networking.k8s.io \"kubermatic\" not found" gateway=kubermatic/kubermatic
time="2026-03-27T11:10:54Z" level=debug msg="   failed to get Gateway, err: Gateway.gateway.networking.k8s.io \"kubermatic\" not found" gateway=kubermatic/kubermatic
time="2026-03-27T11:10:57Z" level=debug msg="   failed to get Gateway, err: Gateway.gateway.networking.k8s.io \"kubermatic\" not found" gateway=kubermatic/kubermatic
time="2026-03-27T11:11:00Z" level=debug msg="   failed to get Gateway, err: Gateway.gateway.networking.k8s.io \"kubermatic\" not found" gateway=kubermatic/kubermatic
[...]
time="2026-03-27T11:27:00Z" level=warning msg="   Timed out waiting for the Gateway to become ready."
time="2026-03-27T11:27:00Z" level=warning msg="   Please check the Gateway and EnvoyProxy Service, and if necessary,"
time="2026-03-27T11:27:00Z" level=warning msg="   reconfigure the envoy-gateway-controller Helm chart. Re-run the installer"
time="2026-03-27T11:27:00Z" level=warning msg="   to apply updated configuration afterwards."
time="2026-03-27T11:27:00Z" level=warning msg="   Gateway kubermatic/kubermatic failed to become ready within 10 minutes: context deadline exceeded"
time="2026-03-27T11:27:00Z" level=info msg="   The main Gateway is ready."
time="2026-03-27T11:27:00Z" level=info msg="   "
time="2026-03-27T11:27:00Z" level=info msg="     Gateway             : kubermatic / kubermatic"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x160 pc=0x294d6df]
[...]
```

Interestingly, everything but the Gateway seems to be deployed:
```
$ k get gatewayclass -A
NAME                       CONTROLLER                     ACCEPTED   AGE
kubermatic-envoy-gateway   gateway.k8c.io/envoy-gateway   True       14m

$ k get envoyproxy -A
NAMESPACE                  NAME                       AGE
envoy-gateway-controller   kubermatic-default-proxy   14m

$ k get gateway -A
No resources found
```

The reason is of course, that the `kubermatic-operator` is missing the required flag.

It's also kinda weird that the `envoy-gateway-controller` Helm chart does not need `migrateGatewayAPI: true` but the `kubermatic-operator` does.

### Expected behavior

kubermatic-installer should detect that `--migrate-gateway-api` was applied but `migrateGatewayAPI: true` is not specified, and exit/fail early and with a helpful error.

### How to reproduce the issue?


### How is your environment configured?

- KKP version: 2.29.5
- Shared or separate master/seed clusters?:

### Provide your KKP manifest here (if applicable)

<!-- Providing an applicable manifest (KubermaticConfiguration, Seed, Cluster or other resources) will help us to reproduce the issue.
Please make sure to redact all secrets (e.g. passwords, URLs...)! -->

<details>

```yaml
# paste manifest here
```

</details>

### What cloud provider are you running on?

<!-- AWS, Azure, DigitalOcean, GCP, Hetzner Cloud, Nutanix, OpenStack, VMware vSphere, Other (e.g. baremetal or non-natively supported provider) -->



### What operating system are you running in your user cluster?

<!-- Ubuntu 20.04, Rocky Linux 8, Flatcar Linux, ... (optional, bug might not be related to user cluster) -->



### Additional information

<!-- Additional information about the bug you're reporting (optional). -->

--- a/cmd/kubermatic-installer/cmd_deploy.go
+++ b/cmd/kubermatic-installer/cmd_deploy.go
@@ -39,6 +39,7 @@ import (
 	"k8c.io/kubermatic/v2/pkg/log"
 	kubernetesprovider "k8c.io/kubermatic/v2/pkg/provider/kubernetes"
 	"k8c.io/kubermatic/v2/pkg/util/flagopts"
+	"k8c.io/kubermatic/v2/pkg/util/yamled"
 	"k8c.io/kubermatic/v2/pkg/version/kubermatic"
 
 	"k8s.io/apimachinery/pkg/util/sets"
@@ -206,6 +207,10 @@ func DeployFunc(logger *logrus.Logger, versions kubermatic.Versions, opt *Deploy
 			return fmt.Errorf("failed to load Helm values: %w", err)
 		}
 
+		if err := validateGatewayAPIMigrationFlags(opt, helmValues); err != nil {
+			return err
+		}
+
 		deployOptions := stack.DeployOptions{
 			HelmClient:                         helmClient,
 			HelmValues:                         helmValues,
@@ -363,6 +368,19 @@ func setupHelmClient(logger *logrus.Logger, opt *DeployOptions) (helm.Client, er
 	}
 
 	return helmClient, nil
+}
+
+func validateGatewayAPIMigrationFlags(opt *DeployOptions, helmValues *yamled.Document) error {
+	if !opt.MigrateGatewayAPI {
+		return nil
+	}
+
+	migrateGatewayAPIInValues, _ := helmValues.GetBool(yamled.Path{"migrateGatewayAPI"})
+	if migrateGatewayAPIInValues {
+		return nil
+	}
+
+	return errors.New("--migrate-gateway-api requires the migrateGatewayAPI Helm value to be set to true")
 }
 
 func setupKubermaticStack(logger *logrus.Logger, args []string, opt *DeployOptions) (stack.Stack, error) {

--- a/cmd/kubermatic-installer/cmd_deploy_test.go
+++ b/cmd/kubermatic-installer/cmd_deploy_test.go
@@ -1,0 +1,80 @@
+/*
+Copyright 2026 The Kubermatic Kubernetes Platform contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"bytes"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"k8c.io/kubermatic/v2/pkg/util/yamled"
+)
+
+func TestValidateGatewayAPIMigrationFlags(t *testing.T) {
+	t.Parallel()
+
+	testCases := []struct {
+		name      string
+		cliFlag   bool
+		helmYAML  string
+		wantError bool
+	}{
+		{
+			name:      "flag disabled and value missing",
+			cliFlag:   false,
+			helmYAML:  "---\n",
+			wantError: false,
+		},
+		{
+			name:      "flag enabled and value set to true",
+			cliFlag:   true,
+			helmYAML:  "migrateGatewayAPI: true\n",
+			wantError: false,
+		},
+		{
+			name:      "flag enabled and value missing",
+			cliFlag:   true,
+			helmYAML:  "---\n",
+			wantError: true,
+		},
+		{
+			name:      "flag enabled and value false",
+			cliFlag:   true,
+			helmYAML:  "migrateGatewayAPI: false\n",
+			wantError: true,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			helmValues, err := yamled.Load(bytes.NewReader([]byte(tc.helmYAML)))
+			require.NoError(t, err)
+
+			err = validateGatewayAPIMigrationFlags(&DeployOptions{MigrateGatewayAPI: tc.cliFlag}, helmValues)
+
+			if tc.wantError {
+				require.ErrorContains(t, err, "--migrate-gateway-api requires the migrateGatewayAPI Helm value to be set to true")
+				return
+			}
+
+			require.NoError(t, err)
+		})
+	}
+}

